src/DataStructures/DataVector.hpp

```diff
@@ -151,407 +181,165 @@ class DataVector {
+template <>
+struct IsVector<DataVector> : std::true_type {};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) atan(
-      const DataVector& t) noexcept {
-    return atan(t.data_);
-  }
+template <>
+struct TransposeFlag<DataVector> : BoolConstant<DataVector::transpose_flag> {};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) atan2(
-      const DataVector& y, const DataVector& x) noexcept {
-    return atan2(~(y.data_), ~(x.data_));
-  }
+template <>
+struct AddTrait<DataVector, DataVector> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) sinh(
-      const DataVector& t) noexcept {
-    return sinh(t.data_);
-  }
+template <>
+struct AddTrait<DataVector, double> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) cosh(
-      const DataVector& t) noexcept {
-    return cosh(t.data_);
-  }
+template <>
+struct AddTrait<double, DataVector> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) tanh(
-      const DataVector& t) noexcept {
-    return tanh(t.data_);
-  }
+template <>
+struct SubTrait<DataVector, DataVector> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) asinh(
-      const DataVector& t) noexcept {
-    return asinh(t.data_);
-  }
+template <>
+struct SubTrait<DataVector, double> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) acosh(
-      const DataVector& t) noexcept {
-    return acosh(t.data_);
-  }
+template <>
+struct SubTrait<double, DataVector> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) atanh(
-      const DataVector& t) noexcept {
-    return atanh(t.data_);
-  }
+template <>
+struct MultTrait<DataVector, DataVector> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) erf(
-      const DataVector& t) noexcept {
-    return erf(t.data_);
-  }
+template <>
+struct MultTrait<DataVector, double> {
+  using Type = DataVector;
+};
 
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) erfc(
-      const DataVector& t) noexcept {
-    return erfc(t.data_);
-  }
-  // @}
+template <>
+struct MultTrait<double, DataVector> {
+  using Type = DataVector;
+};
 
-  /// If less than zero returns zero, otherwise returns one
-  SPECTRE_ALWAYS_INLINE friend decltype(auto) step_function(
-      const DataVector& t) noexcept {
-    return step_function(t.data_);
-  }
+template <>
+struct DivTrait<DataVector, DataVector> {
+  using Type = DataVector;
+};
 
- private:
-  /// \cond HIDDEN_SYMBOLS
-  size_t size_ = 0;
-  InternalStorage_t owned_data_;
-  InternalDataVector_t data_;
-  bool owning_{true};
-  /// \endcond
+template <>
+struct DivTrait<DataVector, double> {
+  using Type = DataVector;
 };
 
-/// Output operator for DataVector
-std::ostream& operator<<(std::ostream& os, const DataVector& d);
+template <typename Operator>
+struct UnaryMapTrait<DataVector, Operator> {
+  using Type = DataVector;
```

Comment:
In #715 we will want a `static_assert(tmpl::list_contains_v<tmpl::list<blaze::Max, blaze::Min, etc>, Operator>, "The Operator is not allowed for ModalVector);` to prevent using things like `sin`, etc. on the ModalVector. Same for the `BinaryMapTrait` (if there are any that are allowed).